File: 22. generate-parentheses/code.java
class Solution {
    List<String> ans = new ArrayList<String>();
    public List<String> generateParenthesis(int n) {
        
        
        rec(n,2*n,0,0,0,"");
        return ans;
    }
    
    void rec(int n,int n1,int i,int open,int close,String s){
        if(i==n1){
            ans.add(s);
            return;
        }
        
        if((open>=close && open+1<=n)){
              s+="(" ;
            rec(n,n1,i+1,open+1,close,s);
            s=s.substring(0,s.length()-1);
        }
        
        if(open>close){
            s+=")";
            rec(n,n1,i+1,open,close+1,s);
        }
        
        return;
    }
}

File: Under 10 mins/22. Generate Parentheses/code.cpp
class Solution {
public:
    set<string> st;
    vector<string> generateParenthesis(int n) {
        fun("", 0, 0, n);

        return vector<string>(st.begin(), st.end());
    }

    void fun(string s, int sa, int e, int n){
        if(e==n) {
            st.insert(s);
            return;}

        if(sa<n)
        fun(s+"(", sa+1, e, n);
        
        if(e<sa)
        fun(s+")", sa, e+1, n);
    }
};
